libs/constrained-generators/src/Constrained/Base.hs

```diff
@@ -1534,7 +1544,10 @@ simplifyPred = \case
     t'@App {} -> Let t' (simplifyBinder b)
     -- Variable or literal
     t' | x :-> p <- b -> simplifyPred $ substitutePred x t' p
-  Exists k b -> Exists k (simplifyBinder b)
+  Exists k b -> case simplifyBinder b of
+    _ :-> TruePred -> TruePred
+    x :-> p | Just t <- pinnedBy x p -> simplifyPred $ substitutePred x t p
+    b' -> Exists k b'
```

Comment:
Maybe put in a comment, with an example that goes down the path that executes this  new code?